### PR TITLE
update UI, add search pTxId

### PR DESCRIPTION
**Pull request by yjqg6666**
# Feature: 
Add search pTxId feature. 
# What changed?
An input box & a search icon are put on the header. 
# How to use?
Type or paste pTxId which may be found from application log and _press the "Enter" key or click on the search icon_, a new page which shows the transaction detail would be open.
# Screenshot
image

### CODE REVIEW

**Comment by yjqg6666** [comment]
pTxId can be obtained from https://github.com/naver/pinpoint/blob/master/doc/per-request_feature_guide.md or PR #6543 (response header).

**Comment by yjqg6666** [comment]
@emeroad Any update for this PR, any thought on this?

**Comment by emeroad** [comment]
@yjqg6666 
I tested this PR.
UI interaction is very simple and nice. 👍

**Comment by dinesh4747** [comment]
@emeroad - Awesome news!!! when this feature will be available, can we expect in next release ?

**Comment by emeroad** [comment]
@yjqg6666 
Waiting for merge. Please wait

**Comment by chkch** [comment]
@emeroad Has it been merged into the master branch?

**Comment by binDongKim** [comment]
@chkch I was supposed to review this sooner but been occupied with other issues. It'll be merged around next week I assume.

**Comment by chkch** [comment]
Good！



---Original---

**Comment by yjqg6666** [comment]
Nothing changed, just rebase onto the new master@d773da8b5 .

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.css [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+.l-search-group-wrap {
+    width: 100%;
+    height: 100%;
+}
+
+.l-search-input {
+    width: 100%;
+    height: 100%;
+    border: 1px solid #d7dde4;
+    padding: 0 5px 0 5px;
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.css]: Adjusting padding will be better for the consistency with other search-input.
`padding: 0 28px 0 10px`

Plus, please add text-overflow ellipsis:
```
text-overflow: ellipsis;
white-space: nowrap;
overflow: hidden;
```

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.css [hunk 2]
```diff
@@ -0,0 +1,35 @@
+:host {
+    display: block;
+    width: 170px;
+    height: 32px;
+    position: relative;
+}
+.l-search-group-wrap {
+    width: 100%;
+    height: 100%;
+}
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.css]: Remove unnecessary style.
```suggestion
```

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.html [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<div class="l-search-group-wrap font-opensans">
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.html]: Remove unnecessary wrapping element.
```suggestion
```

Use other approaches to apply `font-opensans` class such as `HostBinding`.

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.html [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+    <input type="text" placeholder="search pTxId" class="l-search-input"
+           [(ngModel)]="inputValue" (keydown.esc)="onClear()" (keyup.enter)="onSearchId()">
+    <button class="l-search-button fas fa-search" (click)="onSearchId()"></button>
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.html]: Please use `ppSearchInput` directive. It has already taken care of a couple of `Keyboard` events. 
```suggestion
<input 
    type="text"
    ppSearchInput
    placeholder="search pTxId"
    class="l-search-input"
    [useEnter]="searchUseEnter"
    (outSearch)="onSearchId($event)"
    #searchInput>
<button class="l-search-button fas fa-search" (click)="onSearchId(searchInput.value.trim())"></button>
```


You can refer to other use cases in other components.

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    OnDestroy,
+    ChangeDetectionStrategy,
+    Output,
+    EventEmitter
+} from '@angular/core';
+
+import {
+    UrlRouteManagerService
+} from 'app/shared/services';
+import {UrlPath} from "../../../shared/models";
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts]: Use absolute path when importing something.

```suggestion
import { UrlPath } from 'app/shared/models';
```

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    inputValue: string;
+
+    constructor(
+        private urlRouteManagerService: UrlRouteManagerService,
+    ) {}
+
+    ngOnInit() {}
+
+    ngOnDestroy() {}
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts]: Remove unused/unnecessary lifecycle hook.
```suggestion
```

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+import {UrlPath} from "../../../shared/models";
+
+@Component({
+    selector: 'pp-transaction-id-search-container',
+    templateUrl: './transaction-id-search-container.component.html',
+    styleUrls: ['./transaction-id-search-container.component.css'],
+    changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class TransactionIdSearchContainerComponent implements OnInit, OnDestroy {
+    @Output() outSearchId: EventEmitter<{txId: string}> = new EventEmitter();
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts]: Declare `@Output` syntax like below for consistency:
```suggestion
    @Output() outSearchId = new EventEmitter<{txId: string}>();
```

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        if (txId === '') {
+            return;
+        }
+        const spanId = '-1';
+        const collectorAcceptTime = '0';
+        const txInfo = txId.split('^');
+        if (txInfo.length != 3) {
+            return;
+        }
+        const agentId = txInfo[0];
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts]: Give a line-break after `if` statement and variable declaration.
```suggestion
    onSearchId(txId: string): void {
        if (txId === '') {
            return;
        }

        const spanId = '-1';
        const collectorAcceptTime = '0';
        const txInfo = txId.split('^');

        if (txInfo.length !== 3) {
            return;
        }

        const agentId = txInfo[0];
```

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.html [hunk 9]
```diff
@@ -0,0 +1,10 @@
+<input 
+    type="text"
+    ppSearchInput
+    placeholder="search pTxId"
+    class="l-search-input"
+    [useEnter]="searchUseEnter"
+    (outSearch)="onSearchId($event)"
+    #searchInput>
+<button class="l-search-button fas fa-search" (click)="onSearchId(searchInput.value.trim())"></button>
+</div>
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.html]: Remove `</div>`
```suggestion
```

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts [hunk 10]
```diff
@@ -0,0 +1,62 @@
+import {
+    Component,
+    OnInit,
+    OnDestroy,
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts]: Remove unnecessary import
```suggestion
```

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+} from 'app/shared/services';
+import { UrlPath } from 'app/shared/models';
+
+@Component({
+    selector: 'pp-transaction-id-search-container',
+    templateUrl: './transaction-id-search-container.component.html',
+    styleUrls: ['./transaction-id-search-container.component.css'],
+    changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class TransactionIdSearchContainerComponent implements OnInit, OnDestroy {
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts]: Remove unimplemented lifecycle hook
```suggestion
export class TransactionIdSearchContainerComponent implements OnInit {
```

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    @Output() outSearchId = new EventEmitter<{txId: string}>();
+
+    inputValue: string;
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts]: Remove unused property
```suggestion
```

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+    constructor(
+        private urlRouteManagerService: UrlRouteManagerService,
+    ) {}
+
+    ngOnInit() {}
+
+
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts]: Remove unnecessary space
```suggestion
```

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+        const spanId = '-1';
+        const collectorAcceptTime = '0';
+        const txInfo = txId.split('^');
+
+        if (txInfo.length !== 3) {
+            return;
+        }
+
+        const agentId = txInfo[0];
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts:43]: Give a line-break after variable declaration

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+        this.urlRouteManagerService.openPage({
+            path: [
+                UrlPath.TRANSACTION_DETAIL,
+                txId,
+                collectorAcceptTime,
+                agentId,
+                spanId
+            ]
+        });
+        this.onClear();
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts]: Remove unnecessary function call
```suggestion
```

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    onClear() {
+        this.inputValue = '';
+    }
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts]: Remove unnecessary method
```suggestion
```

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts (hunk 6 above)
Comment (yjqg6666) [reply web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts]: Should I also remove "implements OnInit, OnDestroy" ?

Comment (binDongKim) [reply web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts]: You can leave `OnInit` and `ngOnInit()`.

**Comment by yjqg6666** [comment]
Updated. @binDongKim

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+
+@Component({
+    selector: 'pp-transaction-id-search-container',
+    templateUrl: './transaction-id-search-container.component.html',
+    styleUrls: ['./transaction-id-search-container.component.css'],
+    changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class TransactionIdSearchContainerComponent {
+    @Output() outSearchId = new EventEmitter<{txId: string}>();
+
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts:21]: There needs to be a property `searchUseEnter` if you take a look the template.
```suggestion
searchUseEnter = true;
```

File: web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+        const spanId = '-1';
+
+        const collectorAcceptTime = '0';
+
+        const txInfo = txId.split('^');
+
+        if (txInfo.length !== 3) {
+            return;
+        }
```
Comment (binDongKim) [inline web/src/main/angular/src/app/core/components/transaction-id-search/transaction-id-search-container.component.ts:41]: If variable declarations are continuous, no line-break. Sorry if I misled you.
```suggestion
        const spanId = '-1';
        const collectorAcceptTime = '0';
        const txInfo = txId.split('^');

        if (txInfo.length !== 3) {
            return;
        }
```

**Comment by yjqg6666** [comment]
I squashed all the changes from review into one. Thanks a lot for the polishing and consistency check.

**Comment by yjqg6666** [comment]
@binDongKim Since the branch search-txid has already been merge. I made a new PR #6854 for fixing esc-press to clear the search box.

**Comment by dinesh4747** [comment]
@binDongKim @yjqg6666 @emeroad  - when we can expect this? - anywhere around next 2.1.0 release ?

**Comment by yjqg6666** [comment]
@dinesh4747 This feature has been included in Release 2.0.3 and alread been released.

**Comment by dinesh4747** [comment]
@yjqg6666  - Thanks for confirmation, I shall use latest version

With latest change it will automatically impose/construct parent transactionid(pTxID) in app logs and with that I can leverage the search transaction ?

Do we have this feature on exposing pTxid in response headers https://github.com/naver/pinpoint/pull/6543 available in 2.0.3 ?

To expose log in Pinpoint web we need to write some custom logic with controller (// you should implement the logic to retrieve your agent’s logs.) and any pointers on this would help in a great way.

Thanks in anticipation

**Comment by yjqg6666** [comment]
#6543 is not available in official repo. You could fetch the source from naver/pinpoint, checkout tag 2.0.3, add my repo as a new remote, cherry-pick the two commits(5823627b3f2bda127cae642d3950c3cbb041db21 and f7bbac3418f06e163a422cc4c77a6d3a012f98c7), and build your own version. This PR #6789 might also be helpful for self-customized version.

**Comment by zxymd5** [comment]
Can we have more alternatives not just pTxId? For example search by url like '/agent/v1/process/commission/policys'

**Comment by yjqg6666** [comment]
@zxymd5 You can right-click on an edge between two application nodes or single application node and select "filter transaction wizard",  put the url(actually path) into the "URL Pattern" field to search.
image

**Pull request merged**